### ISSUE TITLE
[Feature Request]: add "parent" keyword to `target`

### ISSUE THREAD

**Issue by davestewart** [shentao/vue-global-events#102]
Related to #13 it would be useful to be able to target the immediate parent of a `GlobalEvents` instance, so the parent could manage keypresses within it.

My use case would be to provide different keyboard shortcuts per component (say, an editor component) for different operating systems:

```vue
<RichTextEditor>
  <GlobalEvents v-if="isMac" target="parent" @keyup.meta.b="formatBold" />
  <GlobalEvents v-if="isWin" target="parent" @keyup.ctrl.b="formatBold" />
  ...
<RichTextEditor>
```

**Comment by davestewart** [comment]
Just seen how simple the source code is!

Let me have a look myself...

**Comment by shentao** [comment]
Haha! I know, right? I think this should be quite easy to introduce. We just need to make sure not to break the `target` API.
@posva wdyt?

**Comment by davestewart** [comment]
I'm tinkering about now... not sure I'll be able to write a decent test, but I can update the demo!

**Comment by davestewart** [comment]
So I got it working.

![](http://g.recordit.co/98fGqHHhKA.gif)

It's really nice, as you can add `.stop` and handle it locally, or let it bubble up and be handled in multiple places.

I "think" this will also be perfect for another use case...

I have a bunch of nested UI states, each one can be cancellable via Escape. Using this, I should be able to use one single mechanism to wire it all up.

Happy to push a PR... without tests. Or can work on that if you would prefer.

**Comment by davestewart** [comment]
FYI, the `CTRL+Tab` examples don't seem to work for me on Chrome

**Comment by shentao** [comment]
Nope! I got plenty of work. 😅
Your idea, your PR! A test would be great tho. But we could add it on our own (though this would slow down the merge).
Great job btw!

**Comment by davestewart** [comment]
Ha! Thank you kindly Sir; standing on the shoulders of giants!

**Comment by posva** [comment]
Maybe we could allow an object so we can do `:target="$el"` to target the component. What I don't know is if it will break reactivity by passing a component. Otherwise a string like `parent` or `self` that then triggers a `$parent` inside the component seems fine.

About the test, you can inspire yourself from: https://github.com/shentao/vue-global-events/blob/master/test.js#L215

**Comment by davestewart** [comment]
I like all those options :)

I was about to ask if you had any naming preferences:

```vue
<GlobalEvents :target="$el" @keypress.ctrl.c="copy" />
```
```vue
<GlobalEvents target="self" @keypress.ctrl.c="copy" />
```
```vue
<GlobalEvents target="parent" @keypress.ctrl.c="copy" />
```
```vue
<GlobalEvents target="component" @keypress.ctrl.c="copy" />
```
```vue
<GlobalEvents target="local" @keypress.ctrl.c="copy" />
```
```vue
<GlobalEvents target="this" @keypress.ctrl.c="copy" />
```


And yes, currently using `$parent`. 

Could even pass in a CSS selector seeing as we're at it...

**Comment by shentao** [comment]
I would vote for either `'$parent'` string or passing `$el` or any other ref.

**Comment by davestewart** [comment]
Cool.

Well this handles refs, keywords and selectors:

```js
function getTarget (context) {
  const target = context.target
  if (typeof target === 'string') {
    if (/^(window|document)$/.test(target)) {
      return window[target]
    }
    const $el = context.$parent.$el
    if (target === '$parent') {
      return $el
    }
    return $el.querySelector(target)
  }
  return target
}
```

**Comment by davestewart** [comment]
For consistency, perhaps it should always be `$el`:

Ref:

```
:target="$el"
```

Keyword:

```
target="$el"
```

Selector:

```
target="textarea"
```

**Comment by posva** [comment]
I think that we either pass an object that is an element and leave the string option for window/document or we exclusively allow a string like `parent` to let GlobalEvents attach to `this.$parent` which would be the same as `:target="$el"`. This would avoid adding things that could be done at the parent level:

```js
const $el = context.$parent.$el
    if (target === '$parent') {
      return $el
    }
    return $el.querySelector(target)
```

**Comment by davestewart** [comment]
OK, interestingly... passing references i.e. `$el` or `$refs.textarea` prevents `.stop` from working.

Passing a string such as `"$parent"` and identifying it in the plugin itself does not have this problem.

**Comment by shentao** [comment]
Not exactly sold on the generic selector. :( Although it would work in most cases, I’m afraid it could lead to some issues where the element is not found or has been removed, or there is a duplicate somewhere.
Instead, I would say passing the Vue $ref to an element might be better. There is still the risk that such a DOM element is added and then removed. But this could potentially be handled with a watcher that reruns the contents of `mounted` hook.

**Comment by davestewart** [comment]
> I would say passing the Vue $ref to an element might be better

See previous comment about refs not working with `.stop`

**Comment by shentao** [comment]
Yeah, noticed @davestewart. I wonder what exactly is the reason for that. Feel free to open the PR!

**Comment by davestewart** [comment]
Will do!

**Comment by davestewart** [comment]
Thanks for your time by the way chaps.

I've been mercilessly ignoring my OSS libs for ages. Just can't find the f'ing time.

Props!

**Comment by shentao** [comment]
Oh, man. Tell me about it... 😅 Thanks for getting involved!

**Comment by davestewart** [comment]
OK, got to the bottom of "`.stop` not working".

It's actually because the global event gets called first:

```
:target="$el"
```
```
{text: "ctrl+v (global)", target: textarea, event: KeyboardEvent}
{text: "ctrl+v (local)", target: textarea, event: KeyboardEvent}
```

This is it otherwise:

```
target="$parent"
```
```
{text: "ctrl+v (local)", target: textarea, event: KeyboardEvent}
```

I haven't begun to imagine why...

**Comment by davestewart** [comment]
Looks like adding `.capture` fixes the order:

```vue
<GlobalEvents
  :target="$el"
  @keypress.ctrl.c.capture.stop.prevent="copy"
  @keypress.ctrl.v.capture.stop.prevent="paste"
/>
```

Fugly though.

**Comment by posva** [comment]
@davestewart I tested things locally and I realise it doesn't make sense to use GlobalEvents to attach events to a component or element (or entity) that isn't global and always present (like document, body or window). Regarding your initial use case, I wouldn't use Global Events, I would instead compute the objects of listeners to bind based on the system and bind that with `v-on`

```vue
<RichTextEditor v-on="computedEvents">
  ...
<RichTextEditor>
```

On top of that, the implementation of Global Events, which is based on the assumption that the target it attaches events to is always present and doesn't change while the component GlobalEvents itself is alive, is no longer valid since elements around can get reused but detached from the Dom and reattached later, potentially invalidating events attached by GlobalEvents

**Comment by davestewart** [comment]
Hey @posva ,

Thanks for looking.

To be honest, this PR was a stepping stone to an additional PR which was handling platform-specific keys, which I began to investigate and realised that wouldn't be possible.

Additionally, you closed the [subsequent feature request](https://github.com/vuejs/vue/issues/11302) so I've had to abandon the idea of using your rather useful little library!

However, all is not lost from my end, as taken inspiration from this lib, Vue's own keyboard management, Hotkeys JS and uses [Mousetrap](https://craig.is/killing/mice) to build a really robust Vue hotkeys plugin that is exactly what I need.

```js
keymap () {
  return {
    onHotKeyNew: 'mod+n|global keydown|$window',
    onHotKey: 'any,backspace,del',
    onHotKeyEscape: 'escape',
    onHotKeySelect: 'mod+a',
    onHotKeyDeselect: 'shift+mod+a',
    onHotKeyDetach: 'mod+d',
    onHotKeyClose: 'shift+mod+w',
  }
}
```

Anyway, thanks again.

Cheers,
Dave
